src/iop/lut3d.c

```diff
@@ -0,0 +1,1023 @@
+/*
+    This file is part of darktable,
+    copyright (c) 2016 Chih-Mao Chen
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "bauhaus/bauhaus.h"
+#include "common/imageio_png.h"
+#include "common/colorspaces.h"
+#include "common/colorspaces_inline_conversions.h"
+#include "develop/imageop.h"
+#include "dtgtk/button.h"
+#include "gui/gtk.h"
+#include "iop/iop_api.h"
+
+#include <gtk/gtk.h>
+#include <libgen.h>
+#include <png.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#if defined (_WIN32)
+#include "win/getdelim.h"
+#endif // defined (_WIN32)
+
+DT_MODULE_INTROSPECTION(1, dt_iop_lut3d_params_t)
+
+typedef enum dt_iop_lut3d_colorspace_t
+{
+  DT_IOP_SRGB = 0,
+  DT_IOP_REC709 = 1,
+  DT_IOP_LIN_REC709 = 2,
+  DT_IOP_LIN_PROPHOTORGB = 3,
+} dt_iop_lut3d_colorspace_t;
+
+typedef enum dt_iop_lut3d_interpolation_t
+{
+  DT_IOP_TETRAHEDRAL = 0,
+  DT_IOP_TRILINEAR = 1,
+  DT_IOP_PYRAMID = 2,
+} dt_iop_lut3d_interpolation_t;
+
+typedef void((*dt_interpolation_worker)(float *input, float *output, const float *const clut, const uint8_t level));
+
+typedef struct dt_iop_lut3d_params_t
+{
+  char filepath[512];
+  int colorspace;
+  int interpolation;
+} dt_iop_lut3d_params_t;
+
+typedef struct dt_iop_lut3d_gui_data_t
+{
+//  GtkWidget *lutfolder;
+  GtkWidget *hbox;
+  GtkWidget *filepath;
+  GtkWidget *colorspace;
+  GtkWidget *interpolation;
+} dt_iop_lut3d_gui_data_t;
+
+typedef struct dt_iop_lut3d_data_t
+{
+  dt_iop_lut3d_params_t params;
+  float *clut;  // cube lut pointer
+  uint8_t level; // cube_size
+} dt_iop_lut3d_data_t;
+
+typedef struct dt_iop_lut3d_global_data_t
+{
+  int kernel_lut3d_tetrahedral;
+  int kernel_lut3d_trilinear;
+  int kernel_lut3d_pyramid;
+} dt_iop_lut3d_global_data_t;
+
+const char *name()
+{
+  return _("lut 3D");
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING;
+}
+
+int groups()

```

Comment:
Should probably be:

int default_group()
{
  return IOP_GROUP_COLOR;
}

?

Reply:
yes